source/server-selection/server-selection.rst

```diff
@@ -521,6 +526,15 @@ Eligibility MUST be determined from ``tag_sets`` as follows:
 - If the ``tag_sets`` list is not empty and no tag set in the list matches any
   candidate server, no servers are eligible servers.
 
+hedge
+`````
+
+The read preference ``hedge`` parameter is an object that configures how the
+server will perform hedged reads. It consists of the following keys:
```

Comment:
If we anticipate that the server might introduce additional options (i.e. keys) for `hedge` down the line, do we want to add some language similar to [Unknown Levels and Additional Options for String Based ReadConcerns](https://github.com/mongodb/specifications/blob/master/source/read-write-concern/read-write-concern.rst#unknown-levels-and-additional-options-for-string-based-readconcerns) in the RC/WC spec? I imagine some drivers will naturally want to model the `hedge` option instead of accepting an arbitrary BSON document.

I'll admit the cases aren't entirely analogous because the RC spec is merely prohibiting drivers from accepting new _values_ for a known option/key, while the argument here is whether we should allow drivers to pass new _keys_ to the server. I could be convinced that it's not a problem to allow strongly-typed drivers to model the `hedge` option with known keys and we could leave it unspecified if some drivers decide to just accept an untyped hash/document and ship that to the server. If we permit flexibility to ship arbitrary documents to the server, I would just confirm with the server engineer that we can rely on `mongod` and `mongos` to raise errors for unknown options -- as we'll want to be sure to avoid any case where we let the user specify an option (say, for 4.6+) and have it be silently ignored by a 4.4 server.

Reply:
Tested manually to see the behaviour.

A 4.2 server completely ignores the `hedge` option and does not error when it is encountered:
```
MongoDB Enterprise mongos> db.runCommand({count: 'test', $readPreference: { mode: 'secondary', hedge: {asdf: true} }})
{
	"shards" : {
		
	},
	"n" : 0,
	"ok" : 1,
	"operationTime" : Timestamp(1581060684, 2),
	"$clusterTime" : {
		"clusterTime" : Timestamp(1581060684, 2),
		"signature" : {
			"hash" : BinData(0,"AAAAAAAAAAAAAAAAAAAAAAAAAAA="),
			"keyId" : NumberLong(0)
		}
	}
}
```

So the server will not fail on unknown keys in the read preference document. When running against a 4.3 server, the server also doesn't throw an error when encountering unknown keys in the read preference document, but will fail when encountering an unknown key in the `hedge` document:
```
MongoDB Enterprise mongos> db.runCommand({count: 'test', $readPreference: { mode: 'secondary', hedge: {asdf: true} }})
{
	"ok" : 0,
	"errmsg" : "BSON field 'hedge.asdf' is an unknown field.",
	"code" : 40415,
	"codeName" : "Location40415",
	"operationTime" : Timestamp(1581060865, 1),
	"$clusterTime" : {
		"clusterTime" : Timestamp(1581060865, 1),
		"signature" : {
			"hash" : BinData(0,"AAAAAAAAAAAAAAAAAAAAAAAAAAA="),
			"keyId" : NumberLong(0)
		}
	}
}
```

With this in mind, I would:
* Not require drivers to omit unknown fields in the read preference
* With the driver mantras in mind, we should not check the fields of the `hedge` document for now either. If the server adds additional hedging options in a future release, we will have to check wire versions then before submitting them. For now, relying on a server error seems the best course of action
* Given the impact of submitting a `hedge` option on a 4.2 server (meaning hedging was requested but the server doesn't hedge), I would also not raise an error at all, but I defer to your judgment in this case.

Reply:
> So the server will not fail on unknown keys in the read preference document. When running against a 4.3 server, the server also doesn't throw an error when encountering unknown keys in the read preference document, but will fail when encountering an unknown key in the hedge document:

I'm glad we can rely on 4.4+ to raise errors for unsupported fields in the `hedge` document, as that does justify not having drivers enforce options (unless they need to model those options for strict typing, which they should be free to do).

The lack of validation for read preference fields (across both server versions) is unfortunate. I think we should err on the side of having the client (i.e. driver) raise an error for older wire versions. I believe this would be consistent with what we already do for [`maxStalenessSeconds`](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#maxstalenessseconds):

> During server selection, drivers (but not mongos) MUST raise an error if maxStalenessSeconds is a positive number, and any server's maxWireVersion is less than 5.

"(but not mongos)" suggests that the spec author was aware that mongos will just ignore the field if it is not supported, but a client-side error would pre-empty that from even happening.

Separately, you may want to open a SERVER ticket requesting mongos actually raise errors for unsupported read preference fields.

Reply:
Added a section to require drivers to only send a `hedge` option if maxWireVersion >= 9. I've also opened [SERVER-46093](https://jira.mongodb.org/browse/SERVER-46093) to track checking for unknown read preference options.

Reply:
I don't agree with the rationale for raising a client side error here because silently ignoring the `hedge` option doesn't break any guarantees about the query results. maxStalenessSeconds is different because ignoring maxStalenessSeconds would break the max staleness contract and possibly lead to stale data reads.

I think it's by design that servers ignore unrecognized readPreference options given the wording below:
```
Mongos receiving a query with ``$readPreference`` SHOULD validate the
``mode``, ``tags``, and ``maxStalenessSeconds`` fields according to rules 1 and 2 above,
but SHOULD ignore unrecognized fields for
forward-compatibility rather than throwing an error.
```

But either way is fine with me (client side error or not).

Reply:
Thinking about this again I propose that we do not raise a client-side error. Old servers will ignore the option and that's okay. The benefit of not raising a client-side error is that the server can backport hedged read support if needed and drivers will not need any changes. What do you think?

Reply:
After considering, I've removed the last paragraph requiring drivers not to send a ``hedge`` document if the server does not support it and to throw a client-side error. If the server ticket linked above is resolved to throw errors on unknown read preference options, we can always revisit this.